File: c++/combinationSum.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <cmath>
using namespace std;

typedef vector<vector<int> > vvi;

/*
  dfs to find all combination
  var candidates: candidates
  var size: candidate's length
  var start: because I sort candidates first, dfs() just need to start permutate from "start"
  don't need to start from every element. Then prevent duplicate combinations.
  var now: sum of present numbers
  var target: target
  var save: to save the combination of candidates
  var vvi: the final return vector
*/
void dfs(vector<int> candidates, int size, int start, int now, int target,
         vector<int> &save, vector<vector<int> > &vvi) {
    if (now == target)  {
        for (int i = 0; i < vvi.size(); ++i) {
            if (std::equal(vvi[i].begin(), vvi[i].begin() + vvi[i].size(), save.begin()))
                return;
        }
        vvi.push_back(save);
        return;
    }

    for (int i = start; i < size; ++i) {
        if (now + candidates[i] > target) // above the target value, so the rest(bigger) value is impossible
            return;       // because of this "return" step
        save.push_back(candidates[i]);

        // for Combination Sum, which each number can use "as many times as you want"
        //dfs(candidates, size, i, now + candidates[i], target, save, vvi); // why start from i instead of i+1 ?

        // for Combination Sum II, which each candidate element can only be used "ONCE" !!!!
        dfs(candidates, size, i+1, now + candidates[i], target, save, vvi); 
        save.pop_back();  // this clear step is very important
    }
    return;
}

// my C++ solution use DFS to search all possibility
vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
    vector<vector<int> > vvi;
    vector<int> save;

    // must sort first! then even though there's same element in candidates,
    // it will not produce duplicate combination
    sort(candidates.begin(), candidates.end());
    dfs(candidates, candidates.size(), 0, 0, target, save, vvi);
    return vvi;
}

int main(){


    return 0;
}

File: combinationSum.java
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;
import java.util.Collections;

public class combinationSum {
    public static class Solution {
        // forgot what I wrote here, believe I missed some situations
        public List<List<Integer>> combinationSum0(int[] candidates, int target) {
            List<List<Integer>> list = new ArrayList<List<Integer>>();
            Map<Integer, Integer> map = new HashMap<Integer, Integer>();
            List<Integer> one = new ArrayList<Integer>();
            Map<List<Integer>, Integer> hash = new HashMap<List<Integer>, Integer>();

            for(int i : candidates)
                map.put(i, 1);

            if (map.containsKey(target))
                //list.add(new ArrayList<Integer>(Arrays.asList(target)));
                hash.put(new ArrayList<Integer>(Arrays.asList(target)), 1);
            
            for(int i : candidates) {
                //System.out.println("i: " + i);
                
                for (int j = 1; j <= target/i; j++) { // j = 0, j = 1
                    if ((target - i * j == 0) || map.containsKey(target - i * j)) {
                        int cnt = j;
                        while (cnt > 0) {
                            one.add(i);
                            cnt--;
                        }
                        if (target - i * j != 0)
                            one.add(target - i * j);
                        Collections.sort(one);
                        //System.out.println("one: " + one);
                        hash.put(one, 1);
                        //list.add(one);
                        one = new ArrayList<Integer>();
                    }
                }
            }
            for(List<Integer> key : hash.keySet())
                list.add(key);
            return list;
        }

        // follow my c++ example
        /*
          dfs to find all combination
          var candidates: candidates
          var size: candidate's length
          var start: because I sort candidates first, dfs() just need to start permutate from "start"
          don't need to start from every element. Then prevent duplicate combinations.
          var now: sum of present numbers
          var target: target
          var save: to save the combination of candidates
          var vvi: the final return vector
        */
        public void dfs(int [] candidates, int size, int bgn, int val, int target,
                        List<Integer> one, List<List<Integer>> res) {
            List<Integer> two = new ArrayList<Integer>(one);
            boolean haveit = false;
            if (val == target) {
                for (int i = 0; i < res.size(); i++) { // check if exist already
                    if (res.get(i).size() != two.size()) continue;
                    if (res.get(i).equals(one)) return;
                }
                res.add(two);
                two = new ArrayList<Integer>();
                return;
            }
            for (int i = bgn; i < size; i++) {
                if (val + candidates[i] > target) return;
                two.add(candidates[i]);
                dfs(candidates, size, bgn, val + candidates[i], target, two, res); // use lots of times
                int n = two.size() - 1;
                two.remove(n);
            }
            return;
        }
        
        public List<List<Integer>> combinationSum(int[] candidates, int target) {
            List<List<Integer>> res = new ArrayList<List<Integer>>();
            List<Integer> one = new ArrayList<Integer>();
            Arrays.sort(candidates); // to prevent potential duplicates
            dfs(candidates, candidates.length, 0, 0, target, one, res);
            return res;
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        int [] a = {6, 4, 3, 10, 12};
        System.out.println("");
        for(int i : a)
            System.out.print(i + ", ");
        System.out.println("");

        List<List<Integer>> res = result.combinationSum(a, 28);
        System.out.println("");
        for (int i = 0; i < res.size(); ++i) {
            System.out.println(res.get(i));
        }
        System.out.println("");
    }
}

/*
Time Limit ExceededMore Details // don't know how to make it faster
Last executed input:	[6,4,3,10,12], 28

 */
